Sources/Recycling/ListViews/FavoriteFolders/FavoriteFoldersListView.swift

```diff
@@ -274,23 +275,61 @@ extension FavoriteFoldersListView: UIScrollViewDelegate {
 extension FavoriteFoldersListView: UISearchBarDelegate {
     public func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
         isSearchActive = true
-        footerView.isHidden = true
         searchBar.setShowsCancelButton(true, animated: true)
     }
 
     public func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
-        isSearchActive = false
-        footerView.isHidden = false
-        searchBar.setShowsCancelButton(false, animated: true)
-        delegate?.favoriteFoldersListViewDidCancelSearch(self)
+        // Hide the cancel button if the searchBar is empty.
+        // This will happen if the user removes the text and dismisses the keyboard by dragging.
+        //
+        // Otherwise enable the cancel button.
+        if searchBar.text?.isEmpty ?? true {
+            isSearchActive = false
+            searchBar.setShowsCancelButton(false, animated: true)
+        } else {
+            searchBar.enableCancelButton()
+        }
+    }
+
+    public func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
+        searchBar.resignFirstResponder()
     }
 
     public func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
+        isSearchActive = false
+
         searchBar.text = ""
-        searchBar.endEditing(false)
+        searchBar.setShowsCancelButton(false, animated: true)
+        searchBar.resignFirstResponder()
+
+        delegate?.favoriteFoldersListViewDidCancelSearch(self)
     }
 
     public func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
+        let showCancelButton = searchBar.isFirstResponder || !searchText.isEmpty
+        searchBar.setShowsCancelButton(showCancelButton, animated: true)
+        isSearchActive = !searchText.isEmpty
         delegate?.favoriteFoldersListView(self, didChangeSearchText: searchText)
     }
 }
+
+// MARK: - UISearchBar extension
+
+private extension UISearchBar {
+
+    /// The cancel button is automatically enabled/disabled based on the searchBar's firstResponder status.
+    /// We want to enable it when it's shown, no matter if the searchBar has focus or not.
+    func enableCancelButton() {
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.001, execute: { [weak self] in
+            guard let self = self else { return }
+            for view in self.subviews {
+                for subview in view.subviews {
+                    if let button = subview as? UIButton {
+                        button.isEnabled = true
```

Comment:
This will enable any button inside the search bar to be enabled? I think this is quite dangerous since apple might add another button at any time, or move the cancel button inside something else. 

Is this really something we have to have? Because we don't have an API for doing this so I would really like to avoid it...

Reply:
I think that if we're going to do this, we should do it fully recursively in case the `UIButton` is moved one lever deeper by  at some point:

https://stackoverflow.com/a/46697075
```
    func enableButtons(_ view:UIView) {
        for subView in view.subviews {
            enableButtons(subView)
        }
        if let buttonView = view as? UIButton {
            buttonView.isEnabled = true
        }
    }
```

Reply:
Also, still not digging the need for doing this asynchronously — I'm willing to bet that we'll encounter a race before issues with view hierarchy changes... 🙂